### ISSUE TITLE
TypeScript error for useQuery in v3 react-query

### ISSUE THREAD

**Issue by vrg-success** [tannerlinsley/react-query#1445]
If I specify parameters for useQuery method I get TS error:
```ts
const tariffsQuery = useQuery('tariffs', () => axios.get('/api/tariffs'), { cacheTime: Infinity });
const tariffs = tariffsQuery.data?.data || [];

// Property 'data' does not exist on type 'Promise<never>'.
```
if I delete parameters it's ok
```ts
const tariffsQuery = useQuery('tariffs', () => axios.get('/api/tariffs'));
const tariffs = tariffsQuery.data?.data || [];

// NO-ERRORS
```

**Comment by tannerlinsley** [comment]
Could we get this going in a Codesandbox? That'll help us diagnose things better.

**Comment by vrg-success** [comment]
@tannerlinsley  https://codesandbox.io/s/kind-sinoussi-dypbt?file=/src/App.tsx fds

**Comment by tannerlinsley** [comment]
Looks like the inference isn't working just right for that setup, however, it does work by extracting the function first, which is strange. We'll dig into this.

https://codesandbox.io/s/polished-platform-vg31k?file=/src/App.tsx

**Comment by vrg-success** [comment]
@tannerlinsley  What I gonna do right now? I have a big project, I found reason in useQuery.d.ts
```ts
import { QueryFunction, QueryKey } from '../core/types';
import { UseQueryOptions, UseQueryResult } from './types';
export declare function useQuery<TData = unknown, TError = unknown, TQueryFnData = TData>(options: UseQueryOptions<TData, TError, TQueryFnData>): UseQueryResult<TData, TError>;
export declare function useQuery<TData = unknown, TError = unknown, TQueryFnData = TData>(queryKey: QueryKey, options?: UseQueryOptions<TData, TError, TQueryFnData>): UseQueryResult<TData, TError>;
// Commented TQueryFnData 
export declare function useQuery<TData = unknown, TError = unknown, TQueryFnData = TData>(queryKey: QueryKey, queryFn: QueryFunction</* TQueryFnData | */ TData>, options?: UseQueryOptions<TData, TError, TQueryFnData>): UseQueryResult<TData, TError>;
```
But how can I integrate the type to my project?

**Comment by vrg-success** [comment]
If you have some issue, just create custom.d.ts file and add the code:
```ts

declare module 'react-query' {
  export * from 'react-query/types/index';
  import { QueryFunction, QueryKey } from 'react-query/types/core/types';
  import { UseQueryOptions, UseQueryResult } from 'react-query/types/react/types';
  export declare function useQuery<TData = unknown, TError = unknown, TQueryFnData = TData>(options: UseQueryOptions<TData, TError, TQueryFnData>): UseQueryResult<TData, TError>; // prettier-ignore
  export declare function useQuery<TData = unknown, TError = unknown, TQueryFnData = TData>(queryKey: QueryKey, options?: UseQueryOptions<TData, TError, TQueryFnData>): UseQueryResult<TData, TError>; // prettier-ignore
  export declare function useQuery<TData = unknown, TError = unknown, TQueryFnData = TData>(queryKey: QueryKey, queryFn: QueryFunction</* TQueryFnData | */ TData>, options?: UseQueryOptions<TData, TError, TQueryFnData>): UseQueryResult<TData, TError>; // prettier-ignore
}
```
And don't forget add rules to tsconfig.json

**Comment by tannerlinsley** [comment]
@boschni, it looks like the inference is getting messed up here. My best guess would be that the data inference is attempted by both the queryFn generic and also the config generic (which is possibly overriding the queryFn one): https://github.com/tannerlinsley/react-query/blob/master/src/react/useQuery.ts#L30-L31

I'm not sure how to fix this yet, but maybe you have some ideas.

**Comment by boschni** [comment]
Yeah this also seems to work:
```js
useQuery("tariffs", () => {
  const result = axios.get("/api/tariffs")
  return result
}, { cacheTime: Infinity });

// or:

useQuery('tariffs', () => axios.get<Tariffs>('/api/tariffs'), { cacheTime: Infinity })
```
I'm not 100% sure what is going on, but my guess is that `axios.get` is also trying to infer some types and TS fails to infer everything together.

**Comment by tannerlinsley** [comment]
Interesting. Do you think this is an upstream issue with Axios then?

**Comment by boschni** [comment]
Fixed in version `3.3.1`
